online-shop/src/app/add-product/add-product.component.ts

```diff
@@ -7,35 +7,70 @@ import {
 } from '@angular/forms';
 import { ProductService } from '../all-products/product.service';
 import { Router } from '@angular/router';
+import { Product } from '../classes/product';
 @Component({
   selector: 'app-add-product',
   templateUrl: './add-product.component.html',
 })
-export class AddProductComponent implements OnInit {
-  formGroup: any;
+export class AddProductComponent {
+  formGroup: FormGroup;
   constructor(
     private fb: FormBuilder,
     private productService: ProductService,
     private router: Router
   ) {
     this.formGroup = this.fb.group({
-      name: ['', Validators.required],
-      category: ['', Validators.required],
-      price: ['', Validators.required],
-      description: ['', Validators.required],
-      image: ['', Validators.required],
+      name: [
+        '',
+        [
+          Validators.required,
+          Validators.minLength(3),
+          Validators.maxLength(30),
+        ],
+      ],
+      category: [
+        '',
+        [
+          Validators.required,
+          Validators.minLength(3),
+          Validators.maxLength(30),
+        ],
+      ],
+      price: [
+        '',
+        [Validators.required, Validators.min(1), Validators.max(100000)],
+      ],
+      description: [
+        '',
+        [
+          Validators.required,
+          Validators.minLength(3),
+          Validators.maxLength(300),
+        ],
+      ],
+      image: [
+        '',
+        [
+          Validators.required,
+          Validators.minLength(3),
+          Validators.maxLength(100),
+        ],
+      ],
     });
   }
 
-  ngOnInit(): void {}
-
   onSave() {
+    if (!this.formGroup.valid) {
+      alert('The data you have put in is invalid!');
+    }
     let product = {
+      id: '0',
```

Comment:
you could make the id and quantity optional in your model and get rid of the need to set them to random defaults